### PR TITLE
MM-39711 - Gekidou - Clock functionality [1]

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
This PR adds the Clock setting.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39711

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/hooks/android_back_handler.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {EffectCallback, useEffect} from 'react';
+import {BackHandler} from 'react-native';
+
+import EphemeralStore from '@store/ephemeral_store';
+
+const useAndroidHardwareBackHandler = (componentId: string, callback: EffectCallback) => {
```
Comment (enahum) [inline app/hooks/android_back_handler.tsx:9]: I like this, really do, will you be submitting another PR to replace this in every component that needs it?

File: app/hooks/navigation_button_pressed.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {DependencyList, EffectCallback, useEffect} from 'react';
```
Comment (enahum) [inline app/hooks/navigation_button_pressed.tsx:4]: same question

File: app/screens/settings/display_clock/display_clock.tsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    const onSelectClockPreference = useCallback((clockType: keyof typeof CLOCK_TYPE) => {
+        const militaryTimeFormat = clockType === CLOCK_TYPE.MILITARY;
+        setIsMilitaryTimeFormat(militaryTimeFormat);
+        canSave(hasMilitaryTimeFormat !== militaryTimeFormat);
+    }, [canSave]);
+
+    const close = useCallback(() => popTopScreen(componentId), [componentId]);
+
+    const saveClockDisplayPreference = useCallback(() => {
```
Comment (enahum) [inline app/screens/settings/display_clock/display_clock.tsx:97]: I don't think this needs to be memoized

File: app/screens/settings/display_clock/display_clock.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        setButtons(componentId, buttons);
+    }, [componentId, saveButton]);
+
+    const onSelectClockPreference = useCallback((clockType: keyof typeof CLOCK_TYPE) => {
+        const militaryTimeFormat = clockType === CLOCK_TYPE.MILITARY;
+        setIsMilitaryTimeFormat(militaryTimeFormat);
+        canSave(hasMilitaryTimeFormat !== militaryTimeFormat);
+    }, [canSave]);
+
+    const close = useCallback(() => popTopScreen(componentId), [componentId]);
```
Comment (enahum) [inline app/screens/settings/display_clock/display_clock.tsx:95]: no memoization needed

**Review by enahum — CHANGES_REQUESTED** [review]
need to run i18n-extract

File: app/hooks/android_back_handler.tsx (hunk 1 above)
Comment (avinashlng1080) [reply app/hooks/android_back_handler.tsx:9]: Thank you. Yes, I'll do it.   I have created this ticket to tackle the refactoring: https://mattermost.atlassian.net/browse/MM-45022

File: app/hooks/navigation_button_pressed.tsx (hunk 2 above)
Comment (avinashlng1080) [reply app/hooks/navigation_button_pressed.tsx:4]: Yes

File: app/screens/settings/display_clock/display_clock.tsx (hunk 3 above)
Comment (avinashlng1080) [reply app/screens/settings/display_clock/display_clock.tsx]: corrected

File: app/screens/settings/display_clock/display_clock.tsx (hunk 4 above)
Comment (avinashlng1080) [reply app/screens/settings/display_clock/display_clock.tsx]: corrected

**Review by enahum — CHANGES_REQUESTED** [review]
Still i18n pending

File: app/screens/settings/display_clock/display_clock.tsx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            id: SAVE_CLOCK_BUTTON_ID,
+            enabled: false,
+            showAsAction: 'always' as const,
+            testID: 'settings_display.save.button',
+            color: theme.sidebarHeaderTextColor,
+            text: intl.formatMessage({id: 'settings.display.militaryClock.save', defaultMessage: 'Save'}),
+        };
+    }, [theme.sidebarHeaderTextColor]);
+
+    const canSave = useCallback((enabled: boolean) => {
```
Comment (larkox) [inline app/screens/settings/display_clock/display_clock.tsx:79]: As with other PRs, I think this work better as an effect.

**Review by larkox — APPROVED** [review]
LGTM, just one small thing.

File: app/screens/settings/display_clock/display_clock.tsx (hunk 5 above)
Comment (avinashlng1080) [reply app/screens/settings/display_clock/display_clock.tsx]: corrected

**Review by enahum — APPROVED** [review]
